Enforce __CPROVER_loop_invariant contracts in goto-instrument

SaswatPadhi opened the pull request:
This PR resolves #5882.

It enforces `__CPROVER_loop_invariant` contract (in addition to function contracts) when `--enforce-all-contracts` or `--enforce-contract <fun>` flag is used with `goto-instrument`.

NOTE: The `code_contracts.h` and `code_contracts.cpp` files need a thorough clean up -- the return types seem inconsistent (`void` vs `bool`) for many functions that have similar behavior, and several functions don't use the C++11 range-based for loops etc. I would make a separate PR later for refactoring and cleanup. This PR only adds the invariant checking back into the contracts flags and enables related regression tests.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour. I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

feliperodri commented on regression/contracts/invar_check_01/test.desc:
```diff
@@ -1,11 +1,7 @@
 ^EXIT=0$
 ^SIGNAL=0$
 ^VERIFICATION SUCCESSFUL$
 --
```
Could you add a small explanation about what feature this test case covers? The same goes to all others that you updated.

SaswatPadhi replied:
Thanks for pointing this out! There were comments in the corresponding `.c` files, so I just moved those (with minor edits) to these `.desc` files.

tautschnig commented:
[...]
> I would make a separate PR later for refactoring and cleanup. This PR only adds the invariant checking back into the contracts flags and enables related regression tests.

@SaswatPadhi Please make sure you sync with @feliperodri on this to avoid duplicate work.

feliperodri commented on regression/contracts/invar_loop_constant/test.desc:
```diff
@@ -1,9 +1,10 @@
 --
 --
-Loop invariant checking throws away more information than needed.
+The current implementation of `check_apply_invariants` function
```
What is the exact behavior you expect for this test case?

SaswatPadhi replied:
I clarified the description for this test and added a couple more related tests to demonstrate the expected behavior.

feliperodri reviewed (changes requested):
I'd aslo recommend you to break down the PR into at least two commits: one for the changes in `code_contracts.*` files and another for the updates in the regression tests.

martin-cs commented on regression/contracts/invar_check_break_pass/test.desc:
```diff
@@ -0,0 +1,8 @@
+--enforce-all-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
```
Is there a reason why you don't check the individual VCs as well as the over-all status?

martin-cs commented on regression/contracts/invar_check_continue/test.desc:
```diff
@@ -0,0 +1,8 @@
+--enforce-all-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
```
Again.  I prefer to check VCs as well as it catches bugs that cause VCs to be omitted.

SaswatPadhi replied:
Good idea! I would add VC checks to all tests. Thanks!

martin-cs commented on regression/contracts/invar_check_large/test.desc:
```diff
@@ -0,0 +1,11 @@
+This test checks the invariant contracts on a larger problem --- in this case,
+the partition function of quicksort, applied to a fixed-length array.
+This serves as a stop-gap test until issues to do with quantifiers and
+side-effects in loop invariants are fixed.
```
New line?

martin-cs reviewed (approved):
Seems good.  Comments are only minor improvements.  Thanks for the contribution.

hannes-steffenhagen-diffblue commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -158,7 +158,7 @@ bool code_contractst::has_contract(const irep_idt fun_name)
 }
 
-bool code_contractst::apply_contract(
+bool code_contractst::apply_function_contract(
```
Comment applies to: d20bbe127aa03d8fd8d506c8c43f80871be850ce

Having this refactor take place in the same commit as other changes makes this harder to follow than necessary. This would be easier to follow if, for example, renamings happened in one commit (with an explanation for why the names are better unless obvious), one commit for replacing the loops with ranged-for (don't need an explanation for that beyond "ranged-for is better"), and then finally the actually functional change.

Reason for this is that refactorings should _not_ cause functional changes, whereas any bugfixes or features you add of course _should_ change the functionality, and it makes it a bit hard to see which is which if they're all combined in one changeset.

My preference would be if this commit was split up to that effect, although this is not a hard blocker. Worth considering for the future though!

SaswatPadhi replied:
Thanks Hannes! This is a great suggestion. I just pushed this change.

hannes-steffenhagen-diffblue commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -299,24 +299,16 @@ bool code_contractst::apply_contract(
-  // look at all function calls
-  Forall_goto_program_instructions(ins, goto_function.body)
-    if(ins->is_function_call())
-      apply_contract(goto_function.body, ins);
```
Still d20bbe1

Case in point: In this case it's a bit hard for me to see why this loop was removed. It's combined with a fairly straightforward refactoring of the preceding loop so it's difficult for me see if this was an intentional change or a mistake that snuck in during refactoring.

SaswatPadhi replied:
This particular function was perhaps the old entry point for the old contracts flag. It was used to replace loops with invariant contracts and function calls with function contracts. I removed this loop intentionally because the replacement of calls with contracts now happens in `replace_calls`. You can see a very similar loop in `replace_calls`:

https://github.com/diffblue/cbmc/blob/947fb8fc11d85c21593c3947701babef78b48b45/src/goto-instrument/code_contracts.cpp#L894-L916

hannes-steffenhagen-diffblue reviewed (approved):
I'm ok with the actual changes, clang-format weirdness can be resolved (see my reply to your slack message), would prefer if the first commit was split up further before merging but not a blocker.

SaswatPadhi committed:
Renamed contract application functions

The `code_contracts` function was unused and the function contract
application part of it was improved in `replace_calls`.
This commit:
- removes this old unused functionality,
- renames `code_contracts` to `apply_loop_contract`, and
- renames `apply_contract` to `apply_function_contract`.

SaswatPadhi committed:
Minor refactoring and documentation changes

SaswatPadhi committed:
Enforce __CPROVER_loop_invariant contracts

Loop invariant annotations were being silently ignored when goto-instrument was
invoked with `--enforce-contracts` or `--enforce-contract`.
These flags now enforce both function and loop invariant contracts.

SaswatPadhi committed:
Improved regression tests for invariant contract

- Moved comments from .c files to descriptions in .desc files
- Added new tests to check multiple loops, nested loops, loop-local variables
- Enabled and clarified a regression test that was flagged as a buggy one

SaswatPadhi commented:
Thanks for the review, @martin-cs and @hannes-steffenhagen-diffblue. I think I have addressed all your comments now. Please take another look and let me know if I should make any more changes.

martin-cs reviewed (approved):
Yep; looks good.

martin-cs commented:
I realise this isn't really the right place to ask this but... if you are considering a reworking / rewrite of the code contracts stuff:

1. Is there any chance you could look at / incorporate / test / obsolete the k-induction code in goto-instrument?  There were tests for this somewhere but I don't know if these are enabled.
2. Could it get added to the manual?  I think we are close to having one of the best deductive verification tools for C ( and with @danielsn and @tautschnig 's work on s2n, one of the best used ) but no-one outside would know (I am aware that `goto-analyze` also needs documentation).

SaswatPadhi commented:
Hi @martin-cs,

1. Yes sure, that sounds very related to what I am working on, so I should look at the k-induction code as well. I will take a look at the code and tests next week.

2. Good idea! @feliperodri is maintaining an internal Wiki page documenting the contracts stuff, but we should add these to the public manuals too. Could you please point us to where we should be adding them?

martin-cs commented:
@SaswatPadhi 

1. `src/goto-instrument/k-induction.{h,cpp}` and `regression/k-induction` but it is not routinely run so who knows if it still works!
2. `doc/cprover-manual` would be good.
